### PR TITLE
Mspsds 190 show duplicate products

### PR DESCRIPTION
**Pull request by AnthonyPengelly**

### CODE REVIEW

File: app/assets/javascripts/products.js [hunk 1]
```diff
@@ -8,4 +8,49 @@ $(document).on("turbolinks:load", function() {
...
+}
+
+function buildQuery() {
+    const query = {};
+    const q = $('.new-product-page .search-term:not(#gtin-input)')
+        .map(function() {
+            return $(this).val();
+        })
+        .get()
+        .filter(function(searchTerm) {
```
Comment (broder) [inline app/assets/javascripts/products.js]: Is this filter doing anything?

File: app/assets/javascripts/products.js [hunk 2]
```diff
@@ -8,4 +8,49 @@ $(document).on("turbolinks:load", function() {
...
         adaptiveHeight: true,
     });
+
+    handleSearchInput();
 });
+
+function handleSearchInput() {
+    let debounceTimeout = null;
+    let searchRequest = null;
+    $('.new-product-page .search-term').on('keyup change', function() {
```
Comment (broder) [inline app/assets/javascripts/products.js]: Should we use a library for the debounce functionality?

File: app/controllers/products_controller.rb [hunk 3]
```diff
@@ -72,6 +75,27 @@ def destroy
 
   private
 
+  # If the user supplies a barcode and it matches, then just return that.
+  # Otherwise use the general query param
+  def advanced_product_search
+    products = search_for_gtin
```
Comment (broder) [inline app/controllers/products_controller.rb]: Would something like
```
if params[:gtin].present?
  # search with gtin
else
  search_for_products
```
be neater? Defining `products` multiple times is a bit confusing.

File: app/controllers/products_controller.rb [hunk 4]
```diff
@@ -6,11 +6,14 @@ class ProductsController < ApplicationController
...
-    @products = if params[:q].blank?
-                  Product.paginate(page: params[:page], per_page: 20)
-                else
-                  Product.search(params[:q]).paginate(page: params[:page], per_page: 20).records
-                end
+    @products = search_for_products
+  end
+
+  # GET /products/table
+  def table
```
Comment (broder) [inline app/controllers/products_controller.rb]: Did you have any thoughts about if / how pagination could work with the AJAX that we've got?

File: config/routes.rb [hunk 5]
```diff
@@ -8,6 +8,7 @@
     end
     resources :activities, shallow: true
   end
+  get "/products/table" => "products#table"
```
Comment (broder) [inline config/routes.rb]: Does something like
```
resources :products do
  collection do
    get :table
  end
end
```
work?

File: app/assets/javascripts/products.js (hunk 2 above)
Comment (AnthonyPengelly) [reply app/assets/javascripts/products.js]: We could do, it felt fairly straight forward so I just rolled it myself. The best suggestions for libraries were lodash and underscore which are both quite chunky. Do you think I should go for it anyway?

File: app/assets/javascripts/products.js (hunk 1 above)
Comment (AnthonyPengelly) [reply app/assets/javascripts/products.js]: It's filtering out any undefined/empty values. (I will remove the log though!)

File: app/controllers/products_controller.rb (hunk 4 above)
Comment (AnthonyPengelly) [reply app/controllers/products_controller.rb]: Passing in a page number query param would actually, so we could fairly easily add that on the frontend. However, my gut feel is that it's unlikely that any more than the top few are going to be relevant to the user

File: app/controllers/products_controller.rb (hunk 3 above)
Comment (AnthonyPengelly) [reply app/controllers/products_controller.rb]: I agree that it is currently a bit confusing. However, the functionality is actually a bit different. it's actually more like
```
if params[:gtin].present?
  # try and search with gtin

if params[:q].present? && (there were no results for gtin || !params[:gtin].present?)
  # Search using the query string

else
  # Don't search
```
Does that make sense? I'll have a think about if I can improve the readability though

File: config/routes.rb (hunk 5 above)
Comment (AnthonyPengelly) [reply config/routes.rb]: Works beautifully - thanks!

File: app/assets/javascripts/products.js (hunk 2 above)
Comment (broder) [reply app/assets/javascripts/products.js]: I'm happy to leave it for now. I was hoping that jQuery would have functionality to do this but looks like you need a plugin (https://github.com/cowboy/jquery-throttle-debounce) :/

File: app/controllers/products_controller.rb (hunk 4 above)
Comment (broder) [reply app/controllers/products_controller.rb]: Yeah, hopefully they'll only need the first few results so we should leave it for now

File: app/assets/javascripts/products.js (hunk 1 above)
Comment (broder) [reply app/assets/javascripts/products.js]: Cool!

File: app/controllers/products_controller.rb [hunk 6]
```diff
@@ -78,9 +78,10 @@ def destroy
   # If the user supplies a barcode and it matches, then just return that.
   # Otherwise use the general query param
   def advanced_product_search
-    products = search_for_gtin
-    products = search_for_products if products.blank? && params[:q].present?
-    products
+    gtin_search_results = search_for_gtin if params[:gtin].present?
+    # if there was no GTIN param or there were no results for the GTIN search
+    basic_search_results = search_for_products if gtin_search_results.blank? && params[:q].present?
+    basic_search_results || gtin_search_results || []
```
Comment (broder) [inline app/controllers/products_controller.rb]: Nice!

**Pull request merged**